File: iosApp/iosApp/Views/Content/Grades/GradesViewModel.swift
import Shared
import SwiftUI

extension GradesView {
    @Observable
    class ViewModel: RefreshableViewModel {
        var subjects = [String]()
        var subjectsListTask: Task<(), Never>?
        
        var averages = [[(String, SwiftUI.Color, String)]]()
        var averagesTask: Task<(), Never>?
        
        let repos = RepositoryHelper()
        
        override init() {
            super.init()
            
            self.subjectsListTask = Task {
                for await subjects in self.repos.gradesRepository.getSubjectsListFlow() {
                    self.subjects = subjects
                }
            }
            
            self.averagesTask = Task {
                for await averages in self.repos.gradesRepository.getGeneralAveragesFlow() {
                    self.averages = ["Yearly", "1st semester", "2nd semester"]
                        .enumerated()
                        .map { index, label in
                            ["final", "Proposal"]
                                .enumerated()
                                .compactMap { offset, sublabel in
                                    let index = (index * 2) + offset
                                    guard let average = averages.get(index: Int32(index)) as? Double else { return nil }
                                    
                                    return ("\(offset == 0 ? label + " " : "")\(sublabel)", average.averageToColor(), String(format: "%.2f", average))
                                }
                        }
                        .filter { $0[0].2 != "0.00" || $0[1].2 != "0.00" }
                }
            }
        }
        
        deinit {
            self.subjectsListTask?.cancel()
            self.averagesTask?.cancel()
        }
        
        func refresh() async {
            _ = try? await self.repos.gradesRepository.refresh()
        }
    }
}

extension Double {
    func averageToColor() -> SwiftUI.Color {
        return switch self {
        case 5.75...:
            Color.pink
        case 4.75..<5.75:
            Color.cyan
        case 3.75..<4.75:
            Color.blue
        case 2.75..<3.75:
            Color.green
        case 1.75..<2.75:
            Color.yellow
        case 0.75..<1.75:
            Color.red
        case 0.00..<0.75:
            Color.gray
        default:
            Color.black
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/github/aeoliux/violet/app/content/grades/GradesViewModel.kt
package com.github.aeoliux.violet.app.content.grades

import androidx.compose.material3.MaterialShapes
import androidx.compose.ui.graphics.Color
import androidx.graphics.shapes.RoundedPolygon
import androidx.lifecycle.viewModelScope
import com.github.aeoliux.violet.app.RefreshableViewModel
import com.github.aeoliux.violet.repositories.AlertState
import com.github.aeoliux.violet.repositories.GradesRepository
import com.github.aeoliux.violet.repositories.fill
import com.github.aeoliux.violet.repositories.trimToTheLimit
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn

class GradesViewModel(
    private val gradesRepository: GradesRepository
): RefreshableViewModel() {
    val subjects = this.gradesRepository
        .getSubjectsListFlow()
        .stateIn(viewModelScope, SharingStarted.Lazily, listOf())

    val averagesBySubjectAndSemester = this.gradesRepository
        .getAveragesBySubjectAndSemester()
        .map { averages ->
            averages.forEach { (subject, bySemester) ->
                bySemester.forEachIndexed { semester, average ->
                    println("$subject;${semester + 1} = $average")
                }
            }

            averages
        }
        .stateIn(viewModelScope, SharingStarted.Lazily, linkedMapOf())

    val averages = this.gradesRepository
        .getGeneralAveragesFlow()
        .map { averages ->
            listOf(
                Pair("yearly", 0),
                Pair("1st semester", 2),
                Pair("2nd semester", 4)
            )
                .associateTo(linkedMapOf()) { label ->
                    val pair = listOf(
                        averages[label.second]
                            .let {
                                Triple(
                                    first = "Final",
                                    second = it
                                        .toString()
                                        .trimToTheLimit()
                                        .fill(),
                                    third = it.themeForAverage()
                                )
                            },
                        averages[label.second + 1]
                            .let {
                                Triple(
                                    first = "Proposal",
                                    second = it
                                        .toString()
                                        .trimToTheLimit()
                                        .fill(),
                                    third = it.themeForAverage()
                                )
                            }
                    )

                    Pair(
                        first = label.first,
                        second = pair
                            .takeIf { pair[0].second != "0.00" || pair[1].second != "0.00" }
                    )
                }
                .filterTo(linkedMapOf()) { (label, averages) -> averages != null && averages.isNotEmpty() }
                .mapValues { it.value!! }

        }
        .stateIn(viewModelScope, SharingStarted.Lazily, null)

    fun refresh() = task {
        this.gradesRepository.refresh()
    }
}

fun Double.themeForAverage(): Pair<Color, RoundedPolygon> = when {
    this >= 6 -> Pair(Color.Magenta, MaterialShapes.Burst)
    this >= 4.75 -> Pair(Color.Cyan, MaterialShapes.SoftBurst)
    this >= 3.75 -> Pair(Color.Blue, MaterialShapes.VerySunny)
    this >= 2.75 -> Pair(Color.Green, MaterialShapes.Sunny)
    this >= 1.75 -> Pair(Color.Yellow, MaterialShapes.Cookie4Sided)
    this > 0.0   -> Pair(Color.Red, MaterialShapes.Slanted)
    this == 0.0 -> Pair(Color.Gray, MaterialShapes.Flower)

    else -> Pair(Color.Black, MaterialShapes.Gem)
}